File: mobile/src/api/books.ts
import { apiClient } from './client';

export interface BookSearchResult {
  google_books_id: string;
  title: string;
  authors: string[];
  published_date?: string;
  cover_url?: string;
  description?: string;
  categories: string[];
  page_count?: number;
}

export type BookStatus = 'want_to_read' | 'reading' | 'read' | 'dropped';

export interface UserBookCreate {
  google_books_id: string;
  status: BookStatus;
  format?: 'physical' | 'electronic';
  started_at?: string;
  finished_at?: string;
}

export interface UserBookUpdate {
  status?: BookStatus;
  rating?: number;
  current_page?: number;
  progress_percentage?: number;
  notes?: string;
  started_at?: string;
  finished_at?: string;
}

export interface UserBook {
  id: string;
  status: BookStatus;
  rating?: number;
  current_page: number;
  progress_percentage?: number;
  notes?: string;
  started_at?: string;
  finished_at?: string;
  google_books_id: string;
  title: string;
  authors: string[];
  cover_url?: string;
  page_count?: number;
  published_date?: string;
  description?: string;
  categories: string[];
}

export const booksApi = {
  searchBooks: async (query: string, lang?: string, author?: string): Promise<BookSearchResult[]> => {
    const response = await apiClient.get<BookSearchResult[]>('/books/search', {
      params: { q: query, lang, author },
    });
    return response.data;
  },

  getBookDetails: async (googleBooksId: string): Promise<BookSearchResult> => {
    const response = await apiClient.get<BookSearchResult>(`/books/${googleBooksId}`);
    return response.data;
  },

  getMyBooks: async (status?: BookStatus, limit = 20, offset = 0): Promise<UserBook[]> => {
    const response = await apiClient.get<UserBook[]>('/me/books', {
      params: { status, limit, offset },
    });
    return response.data;
  },

  addBook: async (data: UserBookCreate): Promise<UserBook> => {
    const response = await apiClient.post<UserBook>('/me/books', data);
    return response.data;
  },

  updateBook: async (id: string, data: UserBookUpdate): Promise<UserBook> => {
    const response = await apiClient.patch<UserBook>(`/me/books/${id}`, data);
    return response.data;
  },

  deleteBook: async (id: string): Promise<void> => {
    await apiClient.delete(`/me/books/${id}`);
  },
};

File: backend/app/schemas/books.py
"""
Schemas de libros — validación de datos para búsqueda,
creación y actualización de libros en la colección del usuario.
"""
from pydantic import BaseModel
from datetime import date


class BookSearchResult(BaseModel):
    """Un resultado de búsqueda de Google Books (normalizado)."""
    google_books_id: str
    title: str
    authors: list[str]
    published_date: str | None = None
    cover_url: str | None = None
    description: str | None = None
    categories: list[str] = []
    page_count: int | None = None
    isbn: str | None = None


class UserBookCreate(BaseModel):
    """Datos para añadir un libro a mi colección."""
    google_books_id: str
    status: str = "want_to_read"  # want_to_read, reading, read, dropped
    format: str = "physical"      # physical, electronic
    rating: float | None = None
    started_at: date | None = None
    finished_at: date | None = None


class UserBookUpdate(BaseModel):
    """Datos para actualizar un libro en mi colección."""
    status: str | None = None
    rating: float | None = None
    current_page: int | None = None
    progress_percentage: float | None = None
    format: str | None = None
    notes: str | None = None
    started_at: date | None = None
    finished_at: date | None = None


class UserBookResponse(BaseModel):
    """Un libro en la colección del usuario (con datos del catálogo)."""
    id: str
    status: str
    format: str = "physical"
    rating: float | None = None
    current_page: int = 0
    progress_percentage: float | None = None
    notes: str | None = None
    started_at: date | None = None
    finished_at: date | None = None
    # Datos del libro (del catálogo)
    google_books_id: str
    title: str
    authors: list[str]
    cover_url: str | None = None
    page_count: int | None = None
    published_date: str | None = None
    description: str | None = None
    categories: list[str] = []
    isbn: str | None = None
